### PR TITLE
Language Updates

### PR DESCRIPTION
**Pull request by tgoprince**
This PR introduces a new development version of the plugin, with minor improvements.

Bugs fixed

- Default values for association stereotypes (e.g. bringsAbout, triggers)
- Wrong suggestions for association stereotypes 

New features:

- Added the `<<abstract>>` stereotype
- The `mode` nature on the `restrictedTo` meta-attribute of classes split into `intrinsic-mode` and `extrinsic-mode` to reflect the new UFO taxonomy.

Changes:

- Reversed direction of `<<participation>>`, `<<creation>>`, and `<<termination>>` to match that of gUFO, in which the Endurant class (or Object) is the source and the Event class is the target.
- Refactored some portions of the code to improve readability and minimize duplication

Observations:

- It seems like there is a big change in `SmartModellingController.java`, but it is mostly due to adjusted indentation.

### CODE REVIEW

File: src/main/java/it/unibz/inf/ontouml/vp/utils/SmartColoringUtils.java [hunk 1]
```diff
@@ -26,7 +26,7 @@
 	public static final Color COLOR_FOR_COLLECTIVE = new Color(255, 218, 221);
 	public static final Color COLOR_FOR_COLLECTIVE_DARK = new Color(255, 153, 163);
 	public static final Color COLOR_FOR_EVENT = new Color(252, 252, 212);
-	public static final Color COLOR_FOR_SITUATION = new Color(252, 252, 212);
+	public static final Color COLOR_FOR_SITUATION = new Color(252, 224, 192);
 	public static final Color COLOR_FOR_MODE = new Color(192, 237, 255);
```
Comment (claudenirmf) [inline src/main/java/it/unibz/inf/ontouml/vp/utils/SmartColoringUtils.java:30]: What about we use the color of relators for extrinsic modes as well? they are both relational

File: src/main/java/it/unibz/inf/ontouml/vp/utils/SmartColoringUtils.java [hunk 2]
```diff
@@ -107,7 +107,9 @@ private static Color getColor(IClass _class) {
 			return COLOR_FOR_EVENT;
 		case StereotypesManager.RESTRICTED_TO_SITUATION:
 			return COLOR_FOR_SITUATION;
-		case StereotypesManager.RESTRICTED_TO_MODE:
+		case StereotypesManager.RESTRICTED_TO_INTRINSIC_MODE:
+		case StereotypesManager.RESTRICTED_TO_EXTRINSIC_MODE:
+		case StereotypesManager.RESTRICTED_TO_EXTRINSIC_MODE + " " + StereotypesManager.RESTRICTED_TO_INTRINSIC_MODE:
```
Comment (claudenirmf) [inline src/main/java/it/unibz/inf/ontouml/vp/utils/SmartColoringUtils.java:112]: I'm changing this because sortals must have a unique nature, even modes. For non-sortals, we have an if including quality nature to paint it blue.

File: src/main/java/it/unibz/inf/ontouml/vp/utils/StereotypesManager.java [hunk 3]
```diff
@@ -473,9 +542,10 @@ public static String getDefaultRestrictedTo(String stereotype) {
          case STR_QUALITY:
             return RESTRICTED_TO_QUALITY;
          case STR_MODE:
-            return RESTRICTED_TO_MODE;
+            return RESTRICTED_TO_INTRINSIC_MODE;
```
Comment (claudenirmf) [inline src/main/java/it/unibz/inf/ontouml/vp/utils/StereotypesManager.java:545]: I agree that intrinsic modes should be the default case

File: src/main/java/it/unibz/inf/ontouml/vp/utils/StereotypesManager.java [hunk 4]
```diff
@@ -448,20 +444,93 @@ public static void applyStereotype(IModelElement element, String stereotypeName)
       }
    }
 
+   public static List<String> selectableRestrictedTo(String stereotype) {
```
Comment (claudenirmf) [inline src/main/java/it/unibz/inf/ontouml/vp/utils/StereotypesManager.java:447]: Maybe `availableRestrictionValues()` is a better name (or maybe I'm picky)

**Review by claudenirmf — COMMENTED** [review]
Code reviewed. Going into testing.

**Comment by claudenirmf** [comment]
Does it make sense to enable `isExtensional` to be changed for any class whose restriction is `collective`?

I believe that once a collective class is extensional, this must be the case for subclasses as well. However, it is weird to consider the other way around.

File: src/main/resources/association_constraints.json [hunk 5]
```diff
@@ -378,24 +397,26 @@
   "enumeration:type": [],
   "enumeration:enumeration": [],
   "enumeration:datatype": [],
-  "datatype:kind": [],
```
Comment (claudenirmf) [inline src/main/resources/association_constraints.json:381]: Why did you remove datatype altogether? (I know there no real consequences)

**Review by claudenirmf — COMMENTED** [review]
Double-check constraints

**Comment by claudenirmf** [comment]
I am happy with this PR, just give me your feedback on the comments above so I can generate a new release.

File: src/main/java/it/unibz/inf/ontouml/vp/utils/SmartColoringUtils.java (hunk 1 above)
Comment (tgoprince) [reply src/main/java/it/unibz/inf/ontouml/vp/utils/SmartColoringUtils.java]: Good idea. I made this change and used to opportunity to refactor the smart paint code, making it flexible for future color changes.

File: src/main/java/it/unibz/inf/ontouml/vp/utils/SmartColoringUtils.java (hunk 2 above)
Comment (tgoprince) [reply src/main/java/it/unibz/inf/ontouml/vp/utils/SmartColoringUtils.java]: Are you sure about this claim?

Comment (tgoprince) [reply src/main/java/it/unibz/inf/ontouml/vp/utils/SmartColoringUtils.java]: But regardless, the changes I made to this feature made it flexible to handle any situation. If this is not allowed, we should block users from doing so in the ontological nature dialog.

File: src/main/java/it/unibz/inf/ontouml/vp/utils/StereotypesManager.java (hunk 3 above)
Comment (tgoprince) [reply src/main/java/it/unibz/inf/ontouml/vp/utils/StereotypesManager.java]: Good!

File: src/main/java/it/unibz/inf/ontouml/vp/utils/StereotypesManager.java (hunk 4 above)
Comment (tgoprince) [reply src/main/java/it/unibz/inf/ontouml/vp/utils/StereotypesManager.java]: I renamed it to `possibleRestrictedToValues()`

File: src/main/resources/association_constraints.json (hunk 5 above)
Comment (tgoprince) [reply src/main/resources/association_constraints.json:381]: It was a mistake.

**Comment by tgoprince** [comment]
> Does it make sense to enable `isExtensional` to be changed for any class whose restriction is `collective`?

It definitely does. I changed the condition on the contextual menu to enabled it in such cases.
However, it doesn't work because other stereotypes don't have the isExtensional meta-property.
Could you fix that @claudenirmf?

> I believe that once a collective class is extensional, this must be the case for subclasses as well. However, it is weird to consider the other way around.

I'm not sure. Depends on how we interpret isExtensional when it is false...

**Comment by tgoprince** [comment]
> Double-check constraints

Are you asking me to double-check the constraints or are you telling me you did it? :)

File: src/main/java/it/unibz/inf/ontouml/vp/utils/SmartColoringUtils.java (hunk 2 above)
Comment (claudenirmf) [reply src/main/java/it/unibz/inf/ontouml/vp/utils/SmartColoringUtils.java]: > > Double-check constraints
> 
> Are you asking me to double-check the constraints or are you telling me you did it? :)

I was double-checking the constraints and GitHub required some message to submit my comments. Sorry for the bad naming, no need for you to do anything.

**Comment by claudenirmf** [comment]
> > Double-check constraints
> 
> Are you asking me to double-check the constraints or are you telling me you did it? :)

I was double-checking the constraints and GitHub required some message to submit my comments. Sorry for the bad naming, no need for you to do anything.

**Comment by claudenirmf** [comment]
> > Does it make sense to enable `isExtensional` to be changed for any class whose restriction is `collective`?
> 
> It definitely does. I changed the condition on the contextual menu to enabled it in such cases.
> However, it doesn't work because other stereotypes don't have the isExtensional meta-property.
> Could you fix that @claudenirmf?

Working on it.

> > I believe that once a collective class is extensional, this must be the case for subclasses as well. However, it is weird to consider the other way around.
> 
> I'm not sure. Depends on how we interpret isExtensional when it is false...

Let's think about this constraint, but we may need an enumeration instead of a boolean here to account for three possible value (or we just use `restrictedTo` again)... maybe even for powertypes... 🤔

**Comment by claudenirmf** [comment]
I have updated the code so any stereotypes whose value of `restrictedTo` may be `collective` have the `isExtensional` tagged value. The value of `isExtensional` may be set to `true` iff `collective` is the unique value of `restrictedTo`.

Now updates to `restrictedTo` trigger a verification of `isExtensional`, which is set to `false` in case the class doesn't apply to collectives only. Also, manual changes to the tagged value are prohibited while smart modeling is enabled.

This update accommodates old projects that lacked this tagged value in their stereotypes.

**Comment by claudenirmf** [comment]
Btw, are we using `«characterization»` between intrinsic moments and events? I understand if they could be used as read-only but wanted to check with you.

**Comment by tgoprince** [comment]
> Btw, are we using `«characterization»` between intrinsic moments and events? I understand if they could be used as read-only but wanted to check with you.

Yes, this is allowed by the theory. 

See the paper below by Giancarlo and João. In pager 4, a Trope inheres in an Individual, which can be an Event, a Situation or an Endurant. 

https://www.researchgate.net/publication/336848806_Events_as_Entities_in_Ontology-Driven_Conceptual_Modeling

Also see the object property gufo:inheresIn defined in gUFO: 

http://purl.org/nemo/gufo

**Comment by tgoprince** [comment]
> I have updated the code so any stereotypes whose value of `restrictedTo` may be `collective` have the `isExtensional` tagged value. The value of `isExtensional` may be set to `true` iff `collective` is the unique value of `restrictedTo`.
> 
> Now updates to `restrictedTo` trigger a verification of `isExtensional`, which is set to `false` in case the class doesn't apply to collectives only. Also, manual changes to the tagged value are prohibited while smart modeling is enabled.
> 
> This update accommodates old projects that lacked this tagged value in their stereotypes.

Great solution!

**Comment by tgoprince** [comment]
I just noticed 2 things:

- Disabling the smart-modeling feature does not allow me to access all menu options (only the restrictedTo option).
- When smart-modeling is enabled, we should forbid non-sortals from being marked as `isAbstract = false`

**Comment by tgoprince** [comment]
I fixed the aforementioned issues in my last commit (https://github.com/OntoUML/ontouml-vp-plugin/pull/64/commits/e18f72863e892255d18bb896b71a18bdfccde557).

I think we are ready to merge now.

**Comment by tgoprince** [comment]
Please remember that we have to update the syntax verification on `ontouml-js` to accept situations and abstracts.

**Review by claudenirmf — APPROVED** [review]

**Comment by claudenirmf** [comment]
What do you think about creating a script to use GitHub CLI and automate release generation? We could run something like `mvn release -DpathToReleaseFile=<path>` and create a release with default title and body. I would only need you to create the Maven part and put the plugin version as an environment variable at `./src/main/resources/.properties` (see [this](https://stackoverflow.com/a/3697482/5982716)).

I believe I can write the shell script that makes the actual release and patches the version. This would be mainly for pre-releases of pull requests.

**Comment by claudenirmf** [comment]
I noticed that we have different code styles between Eclipse and IntelliJ. Let's use Google's style instead. I have installed it on Eclipse and enabled auto-formatting on save (see [this](http://www.practicesofmastery.com/post/eclipse-google-java-style-guide/)), you should be able to do the same on IntelliJ.

**Comment by claudenirmf** [comment]
@tgoprince, could you [add the formatter to Maven](https://github.com/coveooss/fmt-maven-plugin) as shows this [tutorial](https://bjurr.com/java-code-formatting-with-google-java-format/)?

I have integrated it to my installation of Eclipse but, beyond this being limited to my computer, it has some annoying side-effects on Eclipse.

**Comment by tgoprince** [comment]
Just to register our decision.

1. I implemented a Maven plugin that automatically formats the code according to Google's Java Format Guidelines upon compiling the code (see https://github.com/OntoUML/ontouml-vp-plugin/pull/64/commits/1c82bbb7ec42a16beda5469cc98d0464d89f1c53)
2. I added a Github action that automatically formats the code, again according to Google's Java Format Guidelines, upon any push to the repository (see https://github.com/OntoUML/ontouml-vp-plugin/pull/64/commits/4864ecdd60a44b84c9d1326149096f5c2bfb5497)

1 is to assist developers in getting the final style while coding
2 is to make sure no unstyled code is committed to the repository

Notice that by pushing commit 2 to Github, I triggered an automatic restyling of all our plugin code (see https://github.com/OntoUML/ontouml-vp-plugin/pull/64/commits/b1e744e5f4debbe50e93c2abf492aa6a02de7eea)

**Comment by claudenirmf** [comment]
After struggling with some issues when deleting associations from diagrams, I have implemented the following strategy for inverting associations (from the example below):

Screen Shot 2020-09-27 at 22 04 07

If you invert an association by making a change for a suggested stereotype, the association ends are kept.

Screen Shot 2020-09-27 at 22 04 26

Screen Shot 2020-09-27 at 22 04 42

If you invert the association through the "Reverse Association" action on the menu, everything is inverted.

Screen Shot 2020-09-27 at 22 05 08

Screen Shot 2020-09-27 at 22 05 30

The code is ugly because I tried to be safe at each step. Please have a look at `Association.invertAssociation()`.

**Comment by claudenirmf** [comment]
Now, SmartModelingController is responsible for keeping the consistency on navigability as well as multiplicity, read-only properties, and aggregation kind. The code now prevents navigability from appearing alongside aggregation kind in the same end (when involving ontouml classes on both sides).

However, SmartModeling runs only when the user changes the stereotype through the context menu. The user is still able to "mess up the navigability" manually through other menus. To prevent this we need to use model listeners as well.

**Comment by claudenirmf** [comment]
There is a known bug in the code, but it is likely to be caused by Visual Paradigm itself and it doesn't seem to break anything.

When hovering the meta-properties menu of an association's source or target, the following exception is thrown if any of the properties are set (i.e., `isReadOnly`, `isDerived`, or `isOrdered`):

```
java.awt.IllegalComponentStateException: component must be showing on the screen to determine its location
```

I believe we can tolerate this exception.

**Comment by tgoprince** [comment]
> Now, SmartModelingController is responsible for keeping the consistency on navigability as well as multiplicity, read-only properties, and aggregation kind. The code now prevents navigability from appearing alongside aggregation kind in the same end (when involving ontouml classes on both sides).
> 
> However, SmartModeling runs only when the user changes the stereotype through the context menu. The user is still able to "mess up the navigability" manually through other menus. To prevent this we need to use model listeners as well.

Sounds like a good feature to consider in the future. I suggest we create an issue for it and merge this PR as it is.

**Comment by tgoprince** [comment]
> There is a known bug in the code, but it is likely to be caused by Visual Paradigm itself and it doesn't seem to break anything.
> 
> When hovering the meta-properties menu of an association's source or target, the following exception is thrown if any of the properties are set (i.e., `isReadOnly`, `isDerived`, or `isOrdered`):
> 
> ```
> java.awt.IllegalComponentStateException: component must be showing on the screen to determine its location
> ```
> 
> I believe we can tolerate this exception.

Were you able to track down which of our methods throws this exception? And if there is any consequence in terms of additional memory or cpu consumption?

**Comment by claudenirmf** [comment]
> Sounds like a good feature to consider in the future. I suggest we create an issue for it and merge this PR as it is.

I have created the issue

> Were you able to track down which of our methods throws this exception? And if there is any consequence in terms of additional memory or cpu consumption?

It happens whenever we hover on these "internal" context menus for the source and target ends meta-properties, regardless with we anything or not. My bet is on some buggy listener on the tool.

I have seen no impact on processor time or memory allocation.

**Comment by claudenirmf** [comment]
I believe that whenever we invert associations we should the association ends where they are, along with their properties (including multiplicity). One important example is when someone defines a parthood relation by hand in the wrong direction and uses the suggested stereotypes menu to set the stereotype:

 - Before applying stereotype: Screen Shot 2020-10-01 at 02 10 47

 - After applying stereotype: Screen Shot 2020-10-01 at 02 11 44

Notice that it is interesting to keep even the multiplicity in place as it carries some information the user made an effort to place there. Keeping the multiplicity in place may be an issue when replacing a «mediation» for a «characterization», for example, but we already have the policy of not overriding user-defined multiplicities.

The main thing we may change from this example is keeping the original reading direction because it must make sense with the name of the association. This will not be in conflict with the real direction of the association, as this is identified by the navigability or the aggregation symbols.

Given everything I said, I am considering removing the dialog warning for inverting associations and the menu button for inverting associations without changing stereotypes.
